crates/proto/src/lib.rs

```diff
@@ -88,3 +132,42 @@ impl Executor for Runtime {
         self.block_on(future)
     }
 }
+
+/// Generic Time for Delay and Timeout.
+// This trait is created to allow to use different types of time systems. It's used in Fuchsia OS, please be mindful when update it.
+pub trait Time {
+    /// Delay type
+    type Delay: 'static + Future<Output = ()> + Send + Unpin;
+
+    /// Return a type that implements `Future` that will wait until the specified duration has
+    /// elapsed.
+    fn delay_for(duration: Duration) -> Self::Delay;
+
+    /// Return a type that implement `Future` to complete before the specified duration has elapsed.
+    fn timeout<F: 'static + Future + Send>(
+        duration: Duration,
+        future: F,
+    ) -> Pin<Box<dyn Future<Output = Result<F::Output, std::io::Error>> + Send>>;
+}
+
+/// New type which is implemented using tokio::time::{Delay, Timeout}
+#[cfg(any(test, feature = "tokio-time"))]
+pub struct TokioTime;
+
+#[cfg(any(test, feature = "tokio-time"))]
+impl Time for TokioTime {
```

Comment:
Maybe we should make this an `async-trait`? Not sure how you feel about that...

for timeout, is there no associated type?

Reply:
I will try it out tomorrow.

Reply:
It seems after I move to async-trait, I don't need to use associate type for either timeout() nor delay_for(). I will update a new version shortly.